### ISSUE TITLE
wangscape for irregular grid

### ISSUE THREAD

**Issue by ilteroi** [Wangscape/Wangscape#170]
hi!

i was thrilled to find this project, seems excellent. though my personal problem is that i'd like to work with an irregular voronoi grid instead of a square grid.

therefore the question: would it be possible to generalize wangscape for any type of grid? i suspect the resulting number of textures would increase significantly, but is it feasible from a coding perspective?

i'd like to give it a try, but first i'd very much appreciate some hints about what i'm getting myself into and where to start.

**Comment by serin-delaunay** [comment]
Hi ilteroi. Thanks for your interest in the project!

Currently Wangscape can only generate rectangular tiles, but we have an issue open (#32) with a plan to support any uniform tile shape with 3, 4, or 6 corners. Not much of it is implemented yet; we have a PR open (#163) which performs some of the preprocessing necessary for arbitrarily-shaped 4-corner tiles.

I can see three ways to support Voronoi tilings in Wangscape. All three require the features in #32 first:
1. If you need strict pixel tessellations for a flat Voronoi tiling, it could be done by making the base/dual[*] tile rearrangement process data-driven, with unlimited numbers of tile shapes, offsets, and a list of adjacencies to make tiles for. This would cost a lot of time to run, and the results might not fit in memory - unless you generated individual tiles rather than full tilesets. I think this is a really bad idea :)
2. Add a fourth geometry type with no base tile, and one high-resolution dual tile in the shape of a triangle. Deform the output tiles to fit the Delaunay triangulation of your Voronoi tiling (it should be a strict triangulation - no squares).
    * Any icons or axis-aligned features would be deformed in the map display.
    * Number of tiles per clique would be T^3, where T is the number of terrains. This is less than the number for square tiles.
    * This may produce edge artefacts depending on how the map is rendered.
    * Manually specify the corner and edge masks for the dual tile, or use function modules to combine central noise modules with border noise modules.
    * Border noise modules would have to be perfectly symmetric (Essentially by reintroducing bug #52 as an option), so that there are no mismatches when juxtaposing the output tiles. This may reduce the visual variety of the final map display - the visual border between Voronoi cells would always intersect the geometric border in the exact middle.
3. Add a fourth geometry type with no base tile, and more than one high-resolution dual tile, all sharing the same equilateral triangular mask. Distinguish the dual tiles by direction arrows along the edges. Apply a direction to each edge of your Voronoi tiling's Delaunay triangulation. When choosing a display for a cell of the triangulation, choose a dual tile which matches the cell's directions.
    * Any icons or axis-aligned features would be deformed in the map display.
    * This may produce edge artefacts depending on how the map is rendered.
    * Manually specify the corner and edge masks for the dual tile, or use function modules to combine central noise modules with border noise modules.
    * This would allow borders to be biased towards either end of the dual tiles without causing mismatches.
    * I think 4 dual tiles (CCC, CCA, CAA, AAA) would be enough for any strict Delaunay triangulation. If you can guarantee that the adjacency graph of the Delaunay triangulation is bipartite, you can use just 2 dual tiles (CCC, AAA) like a hex grid. 2 tiles (CCC, CCA) with reflections providing (AAA, AAC) might also be enough for non-bipartite triangulations.
    * Number of tiles per clique would be D*T^3, where T is the number of terrains and D is the number of dual tiles.

TL;DR Yes we can do it, we need #32 plus a few extra features first, number of tiles won't be a problem, tiles will look deformed in the map display.

[*] "Base tile" here means a Voronoi cell, and "dual tile" is a triangle in the Delaunay triangulation

**Comment by ilteroi** [comment]
wow, thank you for this detailed answer. i need some time to process this, especially option 3 :)

i do think that the texture deformations shouldn't be too bad, assuming the that grid is not "too irregular", ie all  edges have approximately the same length.

so what is your roadmap for future features, in particular implementing #32?

**Comment by serin-delaunay** [comment]
Milestones [2](https://github.com/Wangscape/Wangscape/milestone/2) and [3](https://github.com/Wangscape/Wangscape/milestone/3) are acting as a sort of informal roadmap, but they include a lot of minor issues. At the moment the major features planned are:
* A GUI to replace the command line interface, with an options editor. (@hryniuk is working on this)
* Binary packages for OSX, Arch, Ubuntu, Debian.
* Overhauling alpha calculation (the part of the program which converts noise values to blended pixels), and allowing border textures.
* Supporting non-rectangular geometries. (I'm working on this)

I had written a roadmap and plan for #32 in its issue thread, but they were a little outdated and incomplete. I've updated them now: [roadmap](https://github.com/Wangscape/Wangscape/issues/32#issue-195413505), [plan](https://github.com/Wangscape/Wangscape/issues/32#issuecomment-326616150). It's still subject to change, of course - this is a complicated feature, and I might not have considered every factor while writing the plan.

I think work on supporting Voronoi tilings could start once support for hex tiles is complete.

**Comment by ilteroi** [comment]
well, thanks again, i guess i'll watch this space and concentrate on other issues in the meantime.
